### PR TITLE
[build-script] Factor out a product config object.

### PR DESCRIPTION
**Pull request by ddunbar**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

- This factors out a new configuration object to hold the product specific data
  (like which projects to build). This will grow to include additional options
  for test/install, debug/assertions, etc.
- This also introduces three other small new pieces of infrastructure:
  
  o A facility (--dump-argument-state) to dump the internal argument state as
   JSON, for debugging and testing purposes.
  
  o Some minor new tests using the aforementioned.
  
  o A new NestedNamespace class for defining allowing nested, structured
   arguments to be directly initialized as part of the argument parser logic.
- This is related to SR-237, indirectly, because I plan to lift more of the
  product specific arguments into `build-script` and this makes it easier to
  make things generic on the product.

### CODE REVIEW

**Comment by ddunbar** [comment]
/cc @gribozavr, @rintaro, @karwa

This ended up combining a couple things -- the nested config object support sketched by @rintaro, which I think will help us end up with more structured objects as part of the initial argument parsing, a mode to dump the internal state, and the change itself (an initial product specific config object).

The change to dump the internal state wouldn't be necessary if we factored things so we could just execute the code, but it also has intrinsic value for debugging the script from the command line so I decided it was reasonable for the small amount of code required to dump things as JSON.

My hope is eventually we can get most of the product specific logic to be generic instead of hard coding repeated versions for each product.

File: utils/build-script [hunk 1]
```diff
@@ -1221,42 +1221,22 @@ details of the setups of other systems or automated environments.""")
...
+                       products.LLBuild: "b",
+                       products.SwiftPM: "p"}.get(cls)
+        argnames = []
+        if short_alias is not None:
+            argnames.append("-" + short_alias)
+        argnames.append("--" + cls.product_name())
+        projects_group.add_argument(
+            *argnames, help="enable {}".format(cls.__name__),
+            action=arguments.action.optional_bool,
+            dest="products.{}.build".format(cls.product_name()))
```
Comment (rintaro) [inline utils/build-script]: We cannot use `optional_bool` here. https://github.com/apple/swift/pull/2983#discussion_r67276355

File: utils/swift_build_support/swift_build_support/migration.py [hunk 2]
```diff
@@ -17,10 +17,11 @@
 #
 # ----------------------------------------------------------------------------
 
+import argparse
 import subprocess
 
 
-def parse_args(parser, argv):
+def parse_args(parser, argv, namespace=argparse.Namespace()):
```
Comment (rintaro) [inline utils/swift_build_support/swift_build_support/migration.py]: `namespace=None` is sufficient.

File: utils/swift_build_support/swift_build_support/arguments.py [hunk 3]
```diff
@@ -35,6 +35,75 @@ def _register(registry, name, value):
...
+              self.thing = False
+
+      args = NestedNamespace()
+      args.foo = Foo()
+
+      ...
+
+      parser.add_argument('--foo-thing', dest="foo.thing")
+
+    This class supports accessing the members via both subscription
```
Comment (gribozavr) [inline utils/swift_build_support/swift_build_support/arguments.py]: "subscripting"?

File: utils/swift_build_support/tests/test_build_script.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_product_configuration(self):
+        # Validate which projects are built by default.
+        args = BuildScriptTestCase.get_build_script_argument_state([])
+        self.assertEqual(args.products.cmark.build, True)
+        self.assertEqual(args.products.llvm.build, True)
+        self.assertEqual(args.products.swift.build, True)
+        self.assertEqual(args.products.llbuild.build, False)
+        self.assertEqual(args.products.libdispatch.build, False)
+        self.assertEqual(args.products.swiftpm.build, False)
```
Comment (gribozavr) [inline utils/swift_build_support/tests/test_build_script.py]: Nice!

**Comment by gribozavr** [comment]
LGTM modulo points raised by @rintaro.

**Comment by slavapestov** [comment]
Ping?

**Comment by ddunbar** [comment]
My work refactoring build-script (focused at getting easier to use toolchain products) had to be put on hold given other priorities and since @gribozavr's move. We can probably close this out unless someone else wants to take up the mantle.

**Comment by CodaFi** [comment]
It's been a bit.  Closing this.

**Pull request closed without merging**